backends/bmv2/stf_lexer.py

```diff
@@ -0,0 +1,242 @@
+    t_COLON     = r':'
+    t_COMMA     = r','
+    t_DOT       = r'\.'
+    t_LBRACKET  = r'\['
+    t_RBRACKET  = r'\]'
+    t_LPAREN    = r'\('
+    t_RPAREN    = r'\)'
+    t_EQUAL     = r'='
+    t_EQEQ      = r'=='
+    t_NEQ       = r'!='
+    t_LE        = r'<'
+    t_LEQ       = r'<='
+    t_GT        = r'>'
+    t_GEQ       = r'>='
+    t_SLASH     = r'/'
+
+    # binary constants with ternary (don't care) bits
+    binary_constant     = r'(0[bB][*01]+)'
+
+    hex_prefix          = r'0[xX]'
+    hex_digits          = r'[0-9a-fA-F]'
+    hex_constant_body   = r'(' + hex_digits + r'+)'
+    hex_constant        = r'(' + hex_prefix + hex_constant_body + r')'
+
+    hex_tern            = r'([0-9a-fA-F\*]+)'
+    hex_tern_constant   = r'(' + hex_prefix + hex_tern + r')'
+
+    dec_constant        = r'([0-9]+)'
+
+    identifier          = r'([a-z$A-Z_][a-z$A-Z_0-9]*)'
+
+    @TOKEN(hex_tern_constant)
+    def t_TERN_CONST_HEX(self, t):
+        return t
+
+    @TOKEN(hex_constant)
+    def t_INT_CONST_HEX(self, t):
+        return t
+
+    @TOKEN(binary_constant)
+    def t_INT_CONST_BIN(self, t):
+        return t
+
+    @TOKEN(dec_constant)
+    def t_INT_CONST_DEC(self, t):
+        return t
+
+    # Identifiers in the keyword state should be checked against keywords.
+    # In fact, it should be an error not to find a keyword!!
+    # Throwing that as an error is left as an exercise for next time
+    # when we read the ply manual.
+    @TOKEN(identifier)
+    def t_keyword_ID(self, t):
+        typ = self.keywords_map.get(t.value.lower(), "ID")
+        t.type = typ
+        if typ == 'EXPECT' or typ == 'PACKET':
+            t.lexer.begin('packetdata')
+        else:
+            t.lexer.begin('INITIAL')
+        # print t, "pos:", t.lexpos, "col:", self.lexer.colno
+        return t
+
+    # All identifiers, including keywords are returned as ID outside
+    # the keyword state, except for PACKETS and BYTES (counter types)
+    @TOKEN(identifier)
+    def t_ID(self, t):
+        typ = self.keywords_map.get(t.value.lower(), "ID")
+        if typ == 'BYTES' or typ == 'PACKETS':
+            t.type = typ
+        # print t, "pos:", t.lexpos, "col:", self.lexer.colno
+        return t
+
+    # Discard comments.
+    def t_COMMENT(self, t):
+        r'\#.*$'
+        pass
+
+    # Track line numbers.
+    def t_newline(self, t):
+        r'\n+'
+        t.lexer.lineno += len(t.value)
+        t.lexer.colno = 0
+        t.lexer.begin('keyword')
+
+    # Ignore spaces and tabs.
+    t_ignore = ' \t'
+
+    # Error handling.
+    def t_error(self, t):
+        self._error("Illegal character '%s'" % t.value[0], t)
+
+    # PACKET DATA ------------------------------------------------------------
+
+    @TOKEN(dec_constant)
+    def t_packetdata_DATA_DEC(self, t):
+        return t
+
+    @TOKEN(hex_constant_body)
+    def t_packetdata_DATA_HEX(self, t):
+        return t
+
+    def t_packetdata_DATA_TERN(self, t):
+        r'\*'
```

Comment:
Can someone explain how this is evaluated? I looked at the Ply manual but I still don't understand.

Reply:
if you see a '*' in the string for the packetdata, then the token is a DATA_TERN.